### PR TITLE
add setOutputAsReg

### PR DESCRIPTION
**Pull request by KireinaHoro**
Similar to `Data.setAsReg`, but aware of the port direction that it only sets output ports as register.  This allows a better shorthand like the following:

```scala
val io = new Bundle {
  val masterStream = master(Stream(Bits(4 bits))) setOutputAsReg()
}
```

Using `setAsReg` here would lead to the complaint of `REGISTER DEFINED AS COMPONENT INPUT`.

@Dolu1990 before merging, could you check if `MultiData` is the right place to put this function?  Alternatively, should this be the default behaviour (if direction present, set as register only when is output) for `setAsReg`?

### CODE REVIEW

File: lib/src/main/scala/spinal/lib/bus/amba4/axis/Axi4Stream.scala [hunk 1]
```diff
@@ -36,7 +36,7 @@ object Axi4Stream {
     val keep = (config.useKeep) generate Bits(config.dataWidth bit)
     val last = (config.useLast) generate Bool()
     val dest = (config.useDest) generate UInt(config.destWidth bit)
-    val user = (config.useUser) generate Bits(config.userWidth*config.dataWidth bit)
+    val user = (config.useUser) generate Bits(config.userWidth bit)
```
Comment (Dolu1990) [inline lib/src/main/scala/spinal/lib/bus/amba4/axis/Axi4Stream.scala:39]: Ahh i guess this is an accident ^^

File: core/src/main/scala/spinal/core/HardMap.scala [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package spinal.core
+
+import spinal.idslplugin.Location
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+
+class HardMap extends MultiData {
+  val storage = mutable.LinkedHashMap[NamedType[Data], Data]()
```
Comment (Dolu1990) [inline core/src/main/scala/spinal/core/HardMap.scala:9]: should be in another PR

Comment (KireinaHoro) [reply core/src/main/scala/spinal/core/HardMap.scala:9]: Ah yes, it's an accident that I pushed some other stuff... will do a repush

File: core/src/main/scala/spinal/core/MultiData.scala [hunk 3]
```diff
@@ -103,6 +103,11 @@ abstract class MultiData extends Data {
     this
   }
 
+  def setOutputAsReg(): this.type = {
```
[the remark this answers is not in the document]
Comment (KireinaHoro) [reply core/src/main/scala/spinal/core/MultiData.scala]: Implemented this idea in Data now

**Comment by KireinaHoro** [comment]
@Dolu1990 reimplemented `setOutputAsReg` in Data.  So you think it's a good idea to keep this separate?  (I don't know how to implement it in `Data` and merge it with `setAsReg`)

**Comment by Dolu1990** [comment]
> So you think it's a good idea to keep this separate

Yes :)

All good, thanks !
